### ISSUE TITLE
`ref` declarations with declared types can lie about the type

### ISSUE THREAD

**Issue by vasslitvinov** [chapel-lang/chapel#8947]
Currently the compiler allows a type annotation on a `ref` variable, ex.

```chpl
var asdf: real = 55;
ref fdsa: int = asdf;
```

We should either disallow such annotation or check that the declared type matches the type of the referenced thing. If the latter, the two types must be equal for `ref` and in a supertype relationship for `const ref`.

**Comment by bradcray** [comment]
I'd vote for "allow but check."

**Comment by bradcray** [comment]
This came up on gitter this week, where the following case that I wrote up looks problematic to me:

```chapel
var A: [1..10] int;

ref r1 = A[1];  // OK, r1 will be inferred to be of type 'int'
r1 = 1;

ref r2: int = A[2];  // OK, r2 is declared to be of type 'int'
r2 = 2;

ref r3: real = A[3];  // should be an error: can't make a 'real ref' refer to an 'int'
r3 = 3;

writeln(A);
```

**Comment by vasslitvinov** [comment]
This is less bad than it looks, because the compiler still treats `r3` as an `int`. Regardless, it should generate an error.

**Comment by bradcray** [comment]
> This is less bad than it looks, because the compiler still treats r3 as an int.

I agree that's less bad than it could be, though I also think it's pretty bad in that it prevents a user from being able to rely on the type declarations as documentation, and in cases like this week's gitter conversation, it may cause a user to think they're getting different behavior than they are.  I think we should fix it and am adding a 'user issue' label based on the gitter conversation.

**Comment by bradcray** [comment]
@vasslitvinov : With #23644 merged, what remains to close this issue?  Is it handling class types properly?

In the meeting today, my first reaction was "it's OK with me to put that off", but [re-checking the gitter thread that caused me to add the user issue label a few weeks back](https://matrix.to/#/!PBYDSerrfYujeStENM:gitter.im/$wb2ew90biW6_ngsWhPA80n7exw8T51F0EymfXgOPdGk?via=gitter.im&via=matrix.org&via=gnome.org), I see that it involves classes...

**Comment by vasslitvinov** [comment]
@bradcray this is a good question. To me, these remain open:

* What to do when the declared type of a `ref` is an array type (and it matches the type of the RHS)? Should we add a runtime check? If so, there is an implementation task here. Should there be a compile-time warning?

* Do we want to allow `const ref` aliases to be typed with a parent class relatively to the RHS? If so, it is another implementation task.

**Comment by bradcray** [comment]
Would you be willing to capture those questions as distinct issues that are more specific than this one such that we can discuss each of them in their own thread?  Then we can close this one as being done w.r.t. the OP (which I think it is) and replaced by those ones?

**Comment by vasslitvinov** [comment]
Closing this issue as resolved with #23644.

The above follow-up questions are captured in #23694 and #23695.

**Comment by bradcray** [comment]
Thanks Vass!
